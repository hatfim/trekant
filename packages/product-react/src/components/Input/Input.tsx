import clsx from 'clsx';
import React from 'react';
import FieldWrapper from '../Common/FieldWrapper';
import FormItemWrapper from '../Common/FormItemWrapper';
import IconWrapper from '../Common/IconWrapper';
import RTLWrapper from '../Common/RTLWrapper';

export enum IconPosition {
  Prepend = 'PREPEND',
  End = 'END',
  Start = 'START'
}

type InputProps = React.HTMLProps<HTMLInputElement> & {
  active?: boolean;
  compact?: boolean;
  error?: boolean;
  success?: boolean;
  hint?: string;
  rtl?: boolean;
  formItem?: boolean;
  icon?: React.ReactNode;
  iconPosition?: IconPosition;
};

export const Input: React.FC<InputProps> = ({
  id,
  active,
  compact,
  disabled,
  required,
  error,
  hint,
  icon,
  iconPosition,
  label,
  success,
  value,
  onFocus,
  onBlur,
  rtl,
  formItem,
  ...rest
}) => {
  const [focus, setFocus] = React.useState(false);
  const wrapperClass = clsx('ray-text-field', {
    'ray-text-field--compact': compact,
    'ray-text-field--disabled': disabled,
    'ray-text-field--required': required,
    'ray-text-field--success': success,
    'ray-text-field--error': error,
    'ray-text-field--has-value': value,
    'ray-text-field--active': active || focus,
    'ray-text-field--with-icon-start':
      iconPosition === IconPosition.Start && icon,
    'ray-text-field--with-icon-end': iconPosition === IconPosition.End && icon,
    'ray-text-field--with-prepend':
      iconPosition === IconPosition.Prepend && icon
  });
  const handleFocus = (e: React.FocusEvent<HTMLInputElement>) => {
    setFocus(true);
    onFocus && onFocus(e);
  };
  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    setFocus(false);
    onBlur && onBlur(e);
  };

  return (
    <RTLWrapper renderWrapper={rtl}>
      <FormItemWrapper renderWrapper={formItem}>
        <div className={wrapperClass}>
          <IconWrapper
            renderWrapper={iconPosition === IconPosition.Prepend}
            iconClass="ray-text-field__prepend"
          >
            {icon}
          </IconWrapper>
          <FieldWrapper
            renderWrapper={iconPosition === IconPosition.Prepend}
            fieldClass="ray-text-field__wrapper"
          >
            <input
              id={id}
              className="ray-text-field__input"
              onFocus={handleFocus}
              disabled={disabled}
              onBlur={handleBlur}
              value={value}
              {...rest}
            />
            {label && (
              <label htmlFor={id} className="ray-text-field__label">
                {label}
              </label>
            )}
          </FieldWrapper>
        </div>
        {(formItem || hint) && (
          <div
            className={clsx('ray-form-item__hint', {
              'ray-form-item__hint--error': error,
              'ray-form-item__hint--success': success
            })}
          >
            {hint}
          </div>
        )}
      </FormItemWrapper>
    </RTLWrapper>
  );
};
