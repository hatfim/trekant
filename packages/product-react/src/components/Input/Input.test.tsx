import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { Input } from './Input';

// TODO: we need to consolidate on how we do testing for components.
// we are testing a lot of implementation details (ie: like which classes are rendered)

// describe('Input', () => {
//   let component: ShallowWrapper<Input>;

//   beforeEach(() => {
//     component = shallow<Input>(<Input />);
//   });

//   it('should match snapshot and styles for default props', () => {
//     expect(component).toMatchSnapshot();
//   });
// });
