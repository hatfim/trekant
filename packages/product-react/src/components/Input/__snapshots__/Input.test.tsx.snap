// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match snapshot and styles for default props 1`] = `
<RTLWrapper>
  <FormItemWrapper>
    <div
      className="ray-text-field"
    >
      <IconWrapper
        renderWrapper={false}
      />
      <TextFieldWrapper
        renderWrapper={false}
      >
        <input
          className="ray-text-field__input"
          onBlur={[Function]}
          onFocus={[Function]}
          type="text"
        />
      </TextFieldWrapper>
    </div>
  </FormItemWrapper>
</RTLWrapper>
`;
